File: c++/node.cpp
#include <string>
#include <sstream>
#include "node.hpp"

std::string Header::getString() {
    std::ostringstream ss;
    ss << "<h" << size << ">";
    for (Node* node : children) {
        ss << node->getString();
    }
    ss << "</h" << size << ">";
    return ss.str();
}

std::string Paragraph::getString() {
    std::ostringstream ss;
    ss << "<p>";
    for (Node* node : children) {
        ss << node->getString();
    }
    ss << "</p>\n";
    return ss.str();
}

std::string CodeBlock::getString() {
    std::ostringstream ss;
    ss << "<pre><code>" << text << "</pre></code>\n";
    return ss.str();
}

std::string Image::getString() {
    std::ostringstream ss;
    ss << "<img src=\"" << url << "\" alt=\"" << text << "\" />\n";
    return ss.str();
}

std::string Text::getString() {
    return text;
}

std::string Italic::getString() {
    std::ostringstream ss;
    ss << "<em>";
    for (Node* node : children) {
        ss << node->getString();
    }
    ss << "</em>";
    return ss.str();
}

std::string Bold::getString() {
    std::ostringstream ss;
    ss << "<strong>";
    for (Node* node : children) {
        ss << node->getString();
    }
    ss << "</strong>";
    return ss.str();
}

std::string Code::getString() {
    std::ostringstream ss;
    ss << "<code>" << text << "</code>";
    return ss.str();
}

std::string Link::getString() {
    std::ostringstream ss;
    ss << "<a href=\"" << url << "\">" << text << "</a>";
    return ss.str();
}

File: c#/Node.cs
public interface INode {
    public string GetString();
}

public class Header : INode {
    private int _size;
    private List<INode> _children;

    public Header(int size, List<INode> children) {
        _size = size;
        _children = children;
    }

    public string GetString() {
        string text = "<h" + _size + ">";
        int i = 0;
        foreach (INode child in _children) {
            if (i == 0) {
                text += child.GetString().TrimStart();
            } else {
                text += child.GetString();
            }
            i++;
        }
        return text + "</h" + _size + ">";
    }
}

public class Paragraph : INode {
    private List<INode> _children;

    public Paragraph(List<INode> children) {
        _children = children;
    }

    public string GetString() {
        string text = "<p>";
        foreach (INode child in _children) {
            text += child.GetString();
        }
        return text + "</p>\n";
    }
}

public class CodeBlock : INode {
    private string _text;

    public CodeBlock(string text) {
        _text = text;
    }

    public string GetString() {
        return "<pre><code>" + _text + "</code></pre>\n";
    }
}

public class Image : INode {
    private string _text;
    private string _url;

    public Image(string text, string url) {
        _text = text;
        _url = url;
    }

    public string GetString() {
        return "<img src=\"" + _url + "\" alt=\"" + _text + "\" />\n";
    }
}

public class Text : INode {
    private string _text;

    public Text(string text) {
        _text = text;
    }

    public string GetString() {
        return _text;
    }
}

public class Italic : INode {
    private List<INode> _children;

    public Italic(List<INode> children) {
        _children = children;
    }

    public string GetString() {
        string text = "<em>";
        foreach (INode child in _children) {
            text += child.GetString();
        }
        return text + "</em>";
    }
}

public class Bold : INode {
    private List<INode> _children;

    public Bold(List<INode> children) {
        _children = children;
    }

    public string GetString() {
        string text = "<strong>";
        foreach (INode child in _children) {
            text += child.GetString();
        }
        return text + "</strong>";
    }
}

public class Code : INode {
    private string _text;

    public Code(string text) {
        _text = text;
    }

    public string GetString() {
        return "<code>" + _text + "</code>";
    }
}

public class Link : INode {
    private string _text;
    private string _url;

    public Link(string text, string url) {
        _text = text;
        _url = url;
    }

    public string GetString() {
        return "<a href=\"" + _url + "\">" + _text + "</a>";
    }
}